### ISSUE TITLE
tap-hold-*, tap-next-release: Wrong event order when used quickly

### ISSUE THREAD

**Issue by dset0x** [kmonad/kmonad#558]
Hello,

Given any one of
```
_bs (tap-hold-next 200 bspc lsft)
_bs (tap-next-release bspc lsft)
_bs (tap-hold-next-release 200 bspc lsft)
```

when _rapidly_ typing `_bs j` (where `j` can be any button):

```
Opening KeySink
Registering Uinput device
Opening KeySource
Initiating ioctl grab
Launching process: emitter_proc

--------------------------------------------------------------------------------
Received event: Press <bks>
Running hooks
Block level set to: 1
Registering untimed hook: 2
Registering 200ms hook: 1

--------------------------------------------------------------------------------
Received event: Press <j>
Running hooks
Registering untimed hook: 4
Registering 156ms hook: 3
Storing event, current store: 
 - Press <j>


--------------------------------------------------------------------------------
Received event: Release <bks>
Running hooks
Emitting: Press <bks>
Unblocking input stream, rerunning:
Press <j>

Emitting: Release <bks>

--------------------------------------------------------------------------------
Rerunning event: Press <j>
Running hooks
Registering untimed hook: 5
Emitting: Press <j>

--------------------------------------------------------------------------------
Received event: Release <j>
Running hooks
Emitting: Release <j>
Tried cancelling expired hook: 3
Tried cancelling expired hook: 1
```
```
MappingNotify event, serial 136, synthetic NO, window 0x0,
    request MappingKeyboard, first_keycode 8, count 248

KeyPress event, serial 136, synthetic NO, window 0x9c00001,
    root 0x7b8, subw 0x0, time 212392116, (167,23), root:(1171,939),
    state 0x0, keycode 22 (keysym 0xff08, BackSpace), same_screen YES,
    XLookupString gives 1 bytes: (08) "
    XmbLookupString gives 1 bytes: (08) "
    XFilterEvent returns: False

KeyRelease event, serial 136, synthetic NO, window 0x9c00001,
    root 0x7b8, subw 0x0, time 212392117, (167,23), root:(1171,939),
    state 0x0, keycode 22 (keysym 0xff08, BackSpace), same_screen YES,
    XLookupString gives 1 bytes: (08) "
    XFilterEvent returns: False

KeyPress event, serial 136, synthetic NO, window 0x9c00001,
    root 0x7b8, subw 0x0, time 212392117, (167,23), root:(1171,939),
    state 0x0, keycode 44 (keysym 0x6a, j), same_screen YES,
    XLookupString gives 1 bytes: (6a) "j"
    XmbLookupString gives 1 bytes: (6a) "j"
    XFilterEvent returns: False

KeyRelease event, serial 137, synthetic NO, window 0x9c00001,
    root 0x7b8, subw 0x0, time 212392140, (167,23), root:(1171,939),
    state 0x0, keycode 44 (keysym 0x6a, j), same_screen YES,
    XLookupString gives 1 bytes: (6a) "j"
    XFilterEvent returns: False
```

This works fine if I hold down `_bsp` longer, or if I use `tap-hold` instead.

However, my goal is to make `_bsp` usable to perform `Shift + click`  

Apologies if this is already expressed in another issue - some seemed similar but not quite the same.

**Comment by david-janssen** [comment]
I've also noticed issues with this when trying to quickly perform 2 macros back to back (e.g. I tried to make the spanish `¿¡` work by putting those 2 characters back-to-back, but ran into issues as well). This is stemming from something that needs to be redone, but it's a big fix. It's on my list, but not near the top at the moment. I'll try to remember to post an update here if I ever get to it.

**Comment by dset0x** [comment]
Thanks for the information on this!

I'm seeing also this which is probably the same:

If I have
```
md1 (tap-hold-next-release 180 del  rmet)
nav  (tap-hold-next-release 180 esc (layer-toggle navigation))
```
and I type `P@md1 T@nav R@md1` kmonad will sometimes type`Tdel Tesc` and sometimes `Prmet Tesc Rrmet` (the desired behavior). The latter only happens if I wait 180ms for `rmet` to be emitted.

The workaround I found for this is:
```
md1 (tap-hold-next-release 180 del  rmet
md1 (tap-hold-next 180 del  (around rmet (layer-toggle tmp)))
```
where the `tmp` layer replaces `@nav` with `esc`. Of course removes the ability to use `rmet` with keys in `navigation` but I can work without that for now.

**Comment by dset0x** [comment]
> I've also noticed issues with this when trying to quickly perform 2 macros back to back (e.g. I tried to make the spanish `¿¡` work by putting those 2 characters back-to-back, but ran into issues as well). This is stemming from something that needs to be redone, but it's a big fix. It's on my list, but not near the top at the moment. I'll try to remember to post an update here if I ever get to it.

Reading your comment again after some time, I wonder if the issue you were seeing is the same as mine.

I am simply trying to have a key that acts as a modifier (e.g., shift) when held. When typing fast, it simply doesn't work as intended because kmonad releases it before the next key.

Unfortunately this is extremely frustrating, as half the times I use a modifier key, I end up inputting something else.

**Comment by david-janssen** [comment]
> Reading your comment again after some time, I wonder if the issue you were seeing is the same as mine.

Rereading this stuff again myself, I'm not so sure either. Sadly, I think the most efficient thing right now for me is to focus what time/energy I do have on my main project (getting the IO-shell more solid). Only when that is done does it make sense for me to really turn my mind to digging back into the fine-grained intricacies of the button-model.

My intention is very much to get to this some day.

**Comment by dset0x** [comment]
That's understandable, thanks again!
